### PR TITLE
BerryDex Preview Mode

### PR DESCRIPTION
**Pull request by osherwott**
### Problems that this solves: quickly gathering information on a berry
As adding filters to the berrydex has proven difficult, I have done my best to streamline the process of showing relevant info in the form of showing important parts of the berrydex popup modal within the Berrydex Tab of the Farm. This includes Growth Time and Current Berry Amount, as well as other future implementations, such as Flavors or the upcoming Size and Firmness elements. Filters would still be welcome if someone else wants to help, but this is the best solution I have in the absence of them. My resulting layout was made through the intention of displaying information that was planned to be filterable. A good example of this philosophy is reflected in my addition of Aura highlights.
___
### TASKS: Split PR into smaller ones:
- [x] Berry Aura info #4181 
- [ ] Create row for Parasite status
- [x] Create rows for Wanderers #4422
- [x] CSS variable list setting #4182 
- [x] Change order of information #4266 
- Change how that information is displayed
  - [x] Add "Growth and Berry Stage Length" Row that splits if the time for each is uneven #4265 
  - [ ] Remove aura and parasite information from Description? (maybe have it as a tooltip on the Row?)
- [ ] Add colors and settings to change them #4304 
- [x] Create Berrydex Preview (this PR)
___
_~~To do (also detailed at the bottom)~~ Completed:_
- _clicking again on berry image brings up modal (accessibility)_
- _make sure new modal lines have proper ARIA labels (accessibility)_
- _setting to turn off CSS animations_
- _get feedback/assistance for flexbox/modal sizing_
- _figure out what to do with Aura "multiplier" text for Micle_

### Additions
- **Aura information in Berrydex**
- Tricked out Berrydex with neat CSS stuff (pictured below)
Flavor colors are based upon that of their corresponding condition
https://bulbapedia.bulbagarden.net/wiki/Category:Contest_color_templates
### Changes
- Farm modal **Berry Tab now shows Berrydex preview** with a button for more details on its side
- **Renamed information in the Berrydex Modal** to be clearer to the reader:
Growth time and Berry stage length are named as such
"Total time until Stage" added to Growth Stages section above the times
- **Added information to Berrydex Modal**:
Aura type, if a berry has one
"Aura multiplier at Stage" to Growth Stages section
Aura multiplier numbers, also in the same section
- **Moved rows in the Berrydex modal**:
Hint row to the bottom
Grouped non-time Statistics together underneath Flavor section
### Notes
- Size and Firmness are TBD because **This is to be merged after** #3822
- Farm modal Berry Tab is slightly incompatible with Lux theme (detailed below)

### Berrydex Tab in Farm Modal (default theme)
- Layout upon opening (most likely appears like this even if a player has never encountered a Cheri berry, but that's probably a non-issue)

image

- Comparison with a Berry with an Aura. The colors loop left to right. Note that the "Growth Length" text descriptor is the same in all cases.

image image
- The reorganized detailed Modal. Statistics have been separated into Time and regular Statistics

image image

If a Berry has an unequal growth and time until death, it is displayed here.

image image
___
### Color considerations (compatibility with Solar and other Dark themes)
_Preface: As the default Solar theme already blends with its surroundings, it is assumed that only non-visually impaired people or people who don't prioritize legible themes use Solar, and the application of colors to this theme have been tweaked around that notion._

Berrydex Tab (current vibrancy on the left, illegible example on the right):
image image
image image

Berrydex Modal (for some reason Solar rows are different here, hence its inclusion as an example):
image image
image image

_Note: The Dex Modal is separate from the Tab and can withstand a more vibrant yellow, but I personally recommend the colors matching across both tables. If requested however, it can be done._

The key frames have also been animated to dwell on a legible state (pictured below) for a bit every loop
image
The mentioned frame vs in the middle of the animation:
image image
___
### Help welcome - Nitpicks

- Micle having a unique additive Aura renders the current "Aura Multiplier" text inaccurate
image

- An extra scroll bar appears on the right and the box makes the "More Info" button not immediately noticeable on smaller sizes. Perhaps a flexbox issue? Checking margin sizes is also appreciated, as they were manually selected.
image
image

- Lux theme is too large. The Tab Berrydex is most likely an issue on my part, but it is also too large for the default Modal.
image image

(It also affects other modals)
image

### Help welcome - Accessibility

- Properly labelled html IDs and ARIA compatibility.

- Make it so that clicking on a berry again brings up the modal as normal (if not using a mouse, it can take a while to loop to the "More Info" button with the Tab key)

- Setting to turn off animation (but I also want to make it as unobtrusive as possible to ensure many players can enjoy it regardless of vision impairment)

### CODE REVIEW

**Comment by Aegyo** [comment]
## General thoughts
- I like it, good additions
- The text in Solar screenshots looks kind of... dull? Is it like that everywhere else too? I think at least it would need to be made brighter in the tables with colored background.
- Filters can be implemented separately, got a lot going on here

## Responses to your Help Welcome section:

`Aura Multiplier` -> `Aura Effect`, and list the numbers as x1.5 or +30% like in the aura emitted section of plot tooltips.

Extra toolbar - definitely some css thing. I wouldn't spend too much time trying to figure it out, I can take a look into it when you have finished making changes.

Lux will need some css overrides, I would stick to just the things needed for your additions, and we can open up a separate issue/pr for other Lux issues.

ARIA - Seems difficult to do right, we'd need to test it with actual screen reader software too I think. If you want to add some of this, web.dev has some general guidelines https://web.dev/learn/accessibility/aria-html/ and w3 has more https://www.w3.org/WAI/ARIA/apg/

Clicking again brings up modal - doesn't sound like a problem to add, but it might be difficult to inform players this is a thing? Why don't you have the preview shown on hover (or selection via tab/arrows), and modal open on click (or enter press)

Setting to turn off animation - definitely think this should be a thing, we should do this for all animations really.

**Comment by osherwott** [comment]
- Dull Solar text
Yes, all of it is a soft grey. I don't know how to combat this, but it would definitely need a setting of some sort. I'm considering either having it influence the opacity or switch to a different design, but nothing is decided yet.
- Filters at another time
Agree. This PR is the best I can offer in the absence of them, and I can't promise that their inclusion will come right away, as it's not something I wish to meddle with anytime soon.
- Aura display
That can work.
- Clicking again to bring up modal
I intend this to be part of accessibility accommodations whose function would be clear via ARIA, so informing regular players is not a problem from my perspective. Maybe an extra sentence in the FAQ? (I also want to get this over with, and adding what sounds like a collapsible tab looks like a lot more work, unless someone else helps out. I am not opposed to it, but to be the one doing it!)

File: src/scripts/farming/FarmController.ts [hunk 1]
```diff
@@ -241,5 +241,15 @@ class FarmController {
         return tooltip.join('<br>');
     });
 
+    public static berryClick() {
```
Comment (Vodovik) [inline src/scripts/farming/FarmController.ts:244]:
```suggestion
    
    public static handleBerryDexClick(berryId: number) {
        if ($('#berry-panel').val() === berryId) {
            $('#berryDexModal').modal('show');
        }

        App.game.statistics.selectedBerryID(berryId);
    }
```

File: src/components/farmModal.html [hunk 2]
```diff
@@ -197,14 +199,109 @@
                                      <img src="" width=80px style="padding: 20px 0px;" data-bind="css: { 'pokemon-not-seen': !App.game.farming.unlockedBerries[$data]() }, attr:{ src: FarmController.getBerryImage($data)}">
                                      <!-- ko if: App.game.farming.unlockedBerries[$data]() -->
                                      <span style="bottom: 0; border-bottom-left-radius: 6px; border-bottom-right-radius: 6px" data-bind="text: BerryType[$data]">name</span>
-                                     <a class="overlay" href="#berryDexModal" data-toggle="modal"
+                                     <a id = "berryDexClick" class="overlay" href="#"
```
Comment (Vodovik) [inline src/components/farmModal.html:202]:
```suggestion
                                     <a class="overlay" href='#' data-bind="click: () => {FarmController.handleBerryDexClick($data)}"
```

**Review by Vodovik — COMMENTED** [review]
make the berry modal open when panel already shows this berry

File: src/components/farmModal.html [hunk 3]
```diff
@@ -197,14 +199,109 @@
...
                         </ul>
                     </div>
+                    <div class = "mt-2 flex-grow-1 col-6 col-lg-4" style = "height: 80vh; overflow-y: auto;">
+<!-- ko using: App.game.statistics.selectedBerryID() -->
+                    <div class = "sticky-top">
+                <h5 class="card p-2 modal-title">Berry Tag</h5>
+
+            <div class="modal-header bg-dark text-light" style='justify-content-left: space-around;'>
+                <h5 class="p-0 modal-title text-light" data-bind="text: BerryType[$data]">Berry Statistics</h5>
+                <h5 class="p-0 modal-title text-muted" data-bind="text: `#${(Number($data) + 1).toString().padStart(2,0)}&nbsp;&nbsp;`"></h5>
```
Comment (Vodovik) [inline src/components/farmModal.html:215]:
```suggestion
                <                    <h5 id="berry-panel" class="p-0 modal-title text-muted" data-bind="text: `#${(Number($data) + 1).toString().padStart(2,0)}&nbsp;&nbsp;`, value: $data"></h5>
```

**Comment by osherwott** [comment]
Colors will come in another PR because they need to change with each theme, but this PR is otherwise ready.

**Comment by CypherX** [comment]
It doesn't really look or function good on mobile (I know it scrolls):

image

Some other nitpicks:

Unnecessary things - the three things crossed out with red aren't needed and just add clutter. I know two were copied from the berrydex modal but they aren't really needed here.

Wasted space - margins are quite large and can be reduced (yellow lines), can maybe even fit 4 berries per row. I dislike the big empty space below the grid in the right column but not really sure what to do there.

image

**Comment by osherwott** [comment]
Nice find! I'll try to fix both 👍

**Comment by osherwott** [comment]
I have done my best to fill the extra space, though the one to the right of the berry preview is still there.
I've also changed the icons a bit and tried to match them to each theme.

The extra space under the berry preview will be filled out with wanderers in another PR.

**Comment by osherwott** [comment]
Reopening this to see if my changes will show up (there have been a lot of them)

**Comment by osherwott** [comment]
## Changes
### Included Setting to switch between displays
berrydex mode setting
  - added Aria label to regular berrydex
### List is now of buttons to better match each theme
  - locked berries are disabled buttons, and berries with hints how it in the table on the right
berrydex cerulean
berrydex cerulean hint

**Review by CypherX — APPROVED** [review]
Looks good, thanks

**Pull request merged**